### ISSUE TITLE
Proposal: hook to launch async actions

### ISSUE THREAD

**Issue by Satyam** [reactjs/react-redux#308]
I was looking for where to dispatch an async action to load the initial set of data and it seems that the best places are either the router -by looking for some specific function name within the component to be rendered-  on entering a particular route, or in the component wrapped by `connect`.  Valid as both solutions might be, it seems to me that the data container, being responsible for the data for its wrapped component, would be an even better place.  This becomes more important since React's stateless components deprives them of the lifecycle methods to do it in the component.  Since stateless components are expected to benefit from future optimizations, it would be a shame to miss those if there was a way to keep stateless components stateless.  After all, the data container is already stateful so I came up with this solution:

```js
export default dispatchAsync => Connector => class extends Connector {
  componentDidMount() {
    super.componentDidMount();
    dispatchAsync(this.store.dispatch, this.props, null, this.state.storeState);
  }
  componentWillReceiveProps(nextProps) {
    super.componentWillReceiveProps(nextProps);
    dispatchAsync(this.store.dispatch, nextProps, this.props, this.state.storeState);
  }
};
```

Which is used like this:

```js
import asyncDispatcher from '../utils/asyncDispatcher.js';

const dispatchAsync = (dispatch, nextProps, currentProps, state) => {
  if (isEmpty(state.projects)) dispatch(getAllProjects());
};

export default asyncDispatcher(dispatchAsync)(connect(
  mapStateToProps
)(ProjectList));
```

Just like the mapping functions for `connect`, I define a function `dispatchAsync` that decides whether to dispatch or not.  It receives the `dispatch` function, the future and current properties and the state of the store.

The `asyncDispatcher` function makes a subclass of the `React.Component` returned by `connect` redefining `componentDidMount` and `componentWillReceiveProps`, calling their `super` versions before calling the async dispatcher.  

For `componentDidMount` the async dispatcher gets `null` for `currentProps` as there are none yet.  I am not sure whether passing `{}` instead of `null` would be better.

It would be better if this could be integrated into react-redux, for example, by adding a static function to the `Connect` component.  Then it could be called like this:

```js
export default connect(
  mapStateToProps
)(ProjectList).asyncDispatcher(dispatchAsync);
```

`asyncDispatcher` would get the original class as `this`.

I tried it in a couple of places but made no extensive tests and just client side, no isomorphic setup.  The patch doesn't confuse the React DevTools, the `React` tab in Chrome lists `<Connect(ProjectList) ...` as the component, the patch itself totally invisible.

Something else that would be nice is for the async dispatcher to be able to tell the data container not to bother rendering, for example by returning false (not falsy, === false).  That could set another boolean variable that makes the response to `shouldComponentUpdate`.  After all, if the async dispatcher positively knows the data is not there to be rendered, why waste any time on that?  That flag would be reset in `shouldComponentUpdate` once the return value is calculated. Anyway, this is an optimization, the component should be able to render something without failing when the data is not there.

```js
export default dispatchAsync => Connector => class extends Connector {
  componentDidMount() {
    super.componentDidMount();
    this.noNeedToUpdate = false === dispatchAsync(this.store.dispatch, this.props, null, this.state.storeState);
  }
  //  .....
```

All those arguments would help decide on when to dispatch an action.  A dispatch might always be needed, using some property in `netProps`, it might only be needed if that property in `nextProps` is different from the current or it might be needed if the data for that particular property is not currently cached in the store. I think this is all the information that might be needed to make a decision.

Even if integrated into react-redux, the extra static function would add very few extra bytes to the size of the package, but would not have any other effect on performance if not called or have compatibility issues.

How does it look?  Does it make sense?

**Comment by Satyam** [comment]
This would be the version with the option to `return false` to tell React there is no need to update.

```js
export default dispatchAsync => Connector => class extends Connector {
  componentDidMount() {
    super.componentDidMount();
    this.noNeedToUpdate =
      dispatchAsync(this.store.dispatch, this.props, null, this.state.storeState) === false;
  }
  componentWillReceiveProps(nextProps) {
    super.componentWillReceiveProps(nextProps);
    this.noNeedToUpdate =
      dispatchAsync(this.store.dispatch, nextProps, this.props, this.state.storeState) === false;
  }
  shouldComponentUpdate() {
    const noNeed = this.noNeedToUpdate;
    this.noNeedToUpdate = false;
    return !noNeed && super.shouldComponentUpdate();
  }
};
```

**Comment by Satyam** [comment]
This is the isomorphic version of it:

```js
export default (dispatchAsync, Connector) => class extends Connector {
  componentWillMount() {
    // super.componentWillMount();
    const d = dispatchAsync(this.store.dispatch, this.props, null, this.state.storeState);
    this.noNeedToUpdate = d === false;
    if (d instanceof Promise) {
      this.noNeedToUpdate = true;
      d.then(() => {
        this.noNeedToUpdate = false;
      });
      if (!Array.isArray(this.store.pendingPromises)) this.store.pendingPromises = [];
      this.store.pendingPromises.push(d);
    }
  }
  componentWillReceiveProps(nextProps) {
    super.componentWillReceiveProps(nextProps);
    const d = dispatchAsync(this.store.dispatch, nextProps, this.props, this.state.storeState);
    this.noNeedToUpdate = d === false;
    if (d instanceof Promise) {
      this.noNeedToUpdate = true;
      d.then(() => {
        this.noNeedToUpdate = false;
      });
    }
  }
  shouldComponentUpdate() {
    const noNeed = this.noNeedToUpdate;
    this.noNeedToUpdate = false;
    return !noNeed && super.shouldComponentUpdate();
  }
};
```

As in the version above, the `dispatchAsync` function can return `false` if there is no need to update the screen.

It can also return a Promise.  If that is the case, it assumes the async request operation is not finished yet and sets `noNeedToUpdate` to true so the component doesn't bother updating itself.  It hooks itself into the `then` part of the Promise so it marks that it is now ready to have the screen updated.

To make it isomorphic I have changed the initial hook from `componentDidMount` to `componentWillMount` which is called both for client and server side rendering.

If the value returned from `dispatchAsync` is a Promise, it pushes it into an array in the store in `store.pendingPromises`.

To use it in the server side, you do this:

```js
app.use((req, res, next) => {
  match(
    { routes: clientRoutes, location: req.url },
    (err, redirectLocation, renderProps) => {
      if (err) {
        return res.status(500).end('Internal server error');
      }

      if (!renderProps) {
        return next();
      }

      const initialView = renderToString(
        <Provider store={store}>
          <RouterContext {...renderProps} />
        </Provider>
      );

      Promise.all(store.pendingPromises).then(
        () => {
          const finalState = JSON.stringify(store.getState());
          res.status(200).end(html(initialView, finalState));
        },
        reason => {
          console.error(reason);
          return res.status(500).end(`Internal server error \n${reason}`);
        }
      );
    }
  );
});
```

Basically, after letting it render itself via `renderToString` you call `Promise.all` on `store.pendingPromises`, which has the array of Promises, some resolved some pending.  When they are all solved, then we can read the state of the store and send it to the client.

The async requests would have been performed exactly in the same way, particularly in the same order as in the client, by the component instances with the properties as resolved by `mapStateToProps` and `mergeProps` using not just the routing information but exactly the same props as the client side components would.

**Comment by Satyam** [comment]
My `dispatchAsync` function can return nothing, `false` or a Promise.  I kept it that way for backward compatibility with my previous version, which I am already using.  Otherwise, it agreed that it should always return a Promise, the code would be simpler. 

It would even be simpler if the `store.pendingPromises` array was initialized to an array elsewhere.
